### ISSUE TITLE
Plane.distance(Line) wrong

### ISSUE THREAD

**Issue by smichr** [sympy/sympy#25157]
```python
>>> plane = Plane((0,0,0),(1,0,0),(1,1,0))
>>> plane.distance(Line((x,y,z),(1,1,1)))
0
>>> plane.distance(Line((x,x,y),(1,1,1)))
0
>>>plane.distance(Line((0,0,1),(1,1,1)))
1
```
The symbolic case should not return 0 as we can see there is a non-zero case.

**Comment by smichr** [comment]
The problem is that Plane.distance(Line) starts with 
```python
if self.intersection(o) != []:
            return S.Zero
```
In the case of `plane.intersection(l1)` we have
```python
>>> plane.intersection(l1)
[Point3D(x - z*(x - 1)/(z - 1), y - z*(y - 1)/(z - 1), 0)]
```
Since this is not `[]` it is assumed that an intersection exists and that the smallest distance is zero. But when there is a singularity at `z=1` it means that for that value of `z` there is no intersection (and the distance is 1). So a `Piecewise` could be returned:
```python
Piecewise((1, Eq(z, 1)), (0, True))
```
Some care would have to be taken to handle a case like `Line((a,b,c),(x,y,z))` where the solution would be `Piecewise((1, And(Eq(c, 1), Eq(z, 1))),(0, True))`

**Comment by manns** [comment]
@smichr This would mean that several methods need to be rewritten using Piecewise because results may differ dependent on symbol values. 
I am seeing the following candidates:
- are_concurrent
- equals
- is_coplanar
- is_parallel
- is_perpendicular
- projection (?)

**Comment by smichr** [comment]
Or else NotImplemented for the case where the answer is symbolic in an ambiguous way. This returning of 0 is like wrongly handling `if Eq(a, b): print('true')`: we now raise an error if this doesn't evaluated, otherwise Python semantics would make this True for any non-False value. So for this intersection, we should either return Piecewise if the expression is unevaluated or else raise an error indicating that the symbolic (ambiguous) case is not handled.

I'm not sure how deep we have to go into these other methods; maybe a NotImplemented can just be handled in the `intersection` routine. I haven't (and won't be able to, in the near future) assess this situation.

**Comment by harshanabandara** [comment]
I like to work on this issue. Can someone help me out with the expected answer? Using symbols means that there are a set of lines. In this case there are parallel lines and lines which can intersect the plane. It is kind of confusing. Should it return an symbolic value?

**Comment by harshanabandara** [comment]
@palash018 I think it's the set of lines which go through (1,1,1). So essentially any point can be on this line.

**Comment by palash018** [comment]
@harshanabandara if that is the case then distance shall be 0 the output given in first comment is correct since any point can be on the line these also include points in plane

**Comment by manns** [comment]
@palash018 As I have understood the basic idea, the symbols are considered parameters. Therefore from my point of view, returning just the nearest point for the full set would not be the expected behavior. That means that the result should depend on x, y and z. The answer would have to provide different results for different cases. 

Considering this, @smichr has proposed the usage of `Piecewise` in combination with `NotImplementedError`.  The latter is necessary because with this approach, complex sets may be specified. If you take a Point3D, then you could easily define sets of points that would require handling many or may even infinite cases separately, i.e. a Point3D(x, 1,1) would in reality be a line, but how to handle Point3D(x, y,x**2/y**2)? The same applies for Line3D and Plane objects.

**Comment by smichr** [comment]
yes. The `Line((x,y,z),(1,1,1))` may or may not pass through `Plane((0,0,0),(1,0,0),(1,1,0))` so it is wrong to say that it *does* pass through (and, hence, to return 0). If we aren't going to enumerate with `Piecewise` (the conditions for which a line is parallel to the plane with a given distance) then we should raise a NotImplementedError.
